Block: Improve in file documentation to make it easier to edit content

gziolo opened the pull request:
This PR  tries to improve the experience for the future block implementers by adding inline documentation based on the [Gutenberg Handbook](https://wordpress.org/gutenberg/handbook/). 

Initial suggestion came from @danielbachhuber in https://github.com/wp-cli/scaffold-command/issues/88#issuecomment-347580892:

> Another idea to throw into the mix: Add jsDoc to the scaffolded `edit` and `save` functions that provides an introductory overview to the functions and then links to the corresponding handbook docs.

In addition, I added the following changes:
- Added CSS styles loaded for both editor and fronted.
- Added a note that a recommended way is to generate a block inside a plugin.
- Fixed a few linting errors for generated PHP and JS files.
- Update API to use a new way of setting `supportsHtml`.

### Screenshots

In the editor:

![screen shot 2017-12-21 at 12 22 18](https://user-images.githubusercontent.com/699132/34253863-acc04552-e649-11e7-87f7-fb32810818c4.png)

In the post:

![screen shot 2017-12-21 at 12 22 25](https://user-images.githubusercontent.com/699132/34253864-acdaf1fe-e649-11e7-8920-008f870671ec.png)

gziolo commented on features/scaffold-block.feature:
```diff
@@ -126,6 +126,7 @@ Feature: WordPress block code scaffolding
     Then the {THEME_DIR}/blocks/fight-club.php file should exist
```
2 tests are failing, need to be updated.

2 more tests need to verify if `style.css` file was created.

gziolo commented on templates/block-block-js.mustache:
```diff
@@ -1,33 +1,70 @@
+		 * This represents what the editor will render when the block is used.
+		 * @see https://wordpress.org/gutenberg/handbook/block-edit-save/#edit
+		 *
+		 * @param {Object} props Properties passed to the
```
@param and @return need more love.

gziolo commented on templates/block-block-js.mustache:
```diff
@@ -1,33 +1,70 @@
+		 * which is then serialized by Gutenberg into `post_content`.
+		 * @see https://wordpress.org/gutenberg/handbook/block-edit-save/#save
+		 *
+		 * @return {Object}
```
@return needs more love.

gziolo commented:
@danielbachhuber - how do you regenerate README file? 

I tried `wp scaffold package-readme` but apparently failed with the following:
> Error: 'package-readme' is not a registered subcommand of 'scaffold'. See 'wp help scaffold' for available subcommands.

danielbachhuber commented:
@gziolo Don't worry about it if you can't figure it out. We can do so at the very end.

gziolo commented:
Travis is finally happy, I added a few improvements and addressed all my comments. It needs to have README file regenerated before it gets merged. It's ready for a review.

I want to add support for another template in the follow-up PR.

Another thing I would like to explore is to add a flag for a plugin scaffolding command to optionally generate a block. Does it even make sense?

I want to have all that included in v1.5, so expect my PRs in the middle of January :)

youknowriad commented on templates/block-block-js.mustache:
```diff
@@ -1,33 +1,70 @@
+		 * This is the display title for your block, which can be translated with `i18n` functions.
+		 * The block inserter will show this name.
+		 */
 		title: __( '{{title_ucfirst}}', '{{textdomain}}' ),
```
IIRC the `domain` is not handled yet and I wonder if the i18n module is ready for third-party usage

gziolo replied:
Should I remove `domain` part or leave it as it is? I think it doesn't harm to have 2nd param :)

gziolo replied:
Yes, I should remove the `domain` part from JS file. Related documentation in Gutenberg: https://github.com/WordPress/gutenberg/tree/master/i18n#usage.

> Note that you will not need to specify domain for the strings.

gziolo commented:
> This looks solid to me. I worry a bit about the links to the handbooks in case we change the urls and forget to update here.

The same issue applies to the API. I fix `supportHTML` in this PR.

schlessera commented:
@gziolo To refresh the README file:
1. Install the `wp-cli/scaffold-package-command` package:
```
wp package install wp-cli/scaffold-package-command`
```
2. Run the `wp scaffold package-readme` command in the command's root folder:
```
wp scaffold package-readme . --force
```

gziolo commented:
I removed `textdomain` option as it turned out to be obsolete.

README is regenerated to use the updated PHPDoc. @schlessera thanks for sharing steps to do it 👍

schlessera commented on templates/block-block-js.mustache:
```diff
@@ -1,33 +1,70 @@
 		{{/dashicon}}
+		/**
+		 * Blocks are grouped into categories to help users browse and discover them.
+		 * The core provided categories are `common`, `embed`, `formatting`, `layout` and `widgets`.
```
This seems grammatically off to me. Depending on the intended meaning, I'd suggest either of these two:
- `The provided core categories`
- `The categories provided by core`

schlessera commented on templates/block-block-js.mustache:
```diff
@@ -1,33 +1,70 @@
+		 * @return {Element}       Element to render.
+		 */
 		edit: function( props ) {
 			return el(
```
As this scaffolding serves as an onboarding tool as well, it would be useful to also document what the `el()` does here.

gziolo replied:
👍

gziolo commented:
I added more inline comments to make the generated code better documented.

schlessera commented:
There's still no documentation for the `el()` part, which is rather obscure for newcomers. Would you mind adding a short comment for that as well?
